File: November2021/C++/Number_of_Valid_Words_for_Each_Puzzle.cpp
class Solution {
public:
    vector<int> findNumOfValidWords(vector<string>& words, vector<string>& puzzles) {
            unordered_map<char,vector<int>> dict;
            for(int i=0;i<26;i++)
            {
                vector<int> v;
                dict[(char)(i+'a')]=v;
            }
        
        for(auto& word:words)
        {
            int mask{};
            
            for(auto& ch:word)
            {
                int bit =ch-'a';
                mask=mask|1<<bit;
            }
            
            vector<int> rep(26,0);
            for(auto& ch:word)
            {
                
                if(!rep[ch-'a'])
                {
                    dict[ch].push_back(mask);
                    rep[ch-'a']=1;
                }
            }
        }
        vector<int> ans;
        for(auto& puzzle :puzzles)
        {
            int cnt{};
            char firstchar=puzzle[0];
            int mask{};
            
            for(auto& ch:puzzle)
            {
                int bit=ch-'a';
                mask=mask|1<<bit;
            }
            
            // vector<int> res=dict[firstchar];
            for(auto& wrd:dict[firstchar])
            {
                if((mask&wrd)==wrd)
                    cnt++;
            }
            ans.push_back(cnt);
                
        }
        return ans;
        
    }
};

File: November2021/NumberOfValidWordsForEachPuzzle.java
class Solution {
	public List<Integer> findNumOfValidWords(String[] words, String[] puzzles) {
		// English Character
		Map<Integer, List<Integer>> bucket = new HashMap<>();
		for (int i = 0; i < 26; i++) {
			bucket.put(i, new ArrayList<>());
		}
		for (String word : words) {
			int temp = getBitMask(word);
			//
			for (int i = 0; i < 26; i++) {
				if ((temp & (1 << i)) != 0) {
					bucket.get(i).add(temp);
				}
			}
		}
		List<Integer> res = new ArrayList<>();
		for (String p : puzzles) {
			int puzzleMask =  getBitMask(p);
			int c = p.charAt(0) - 'a';
			int count = 0;
			// searching for those words that have first puzzle char
			for (int wordMask : bucket.get(c)) {
				if ((wordMask & puzzleMask) == wordMask) {
					count++;
				}
			}
			res.add(count);
		}
		return res;
	}

	private int getBitMask(String s) {
		int mask = 0;

		for (int i = 0; i < s.length(); ++i) {
			char c = s.charAt(i);
			mask |= getBitMask(c);
		}

		return mask;
	}

	private int getBitMask(char c) {
		int bit = (c - 'a');
		return (1 << bit);
	}
}